File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/voicefilter/dsp/Fft512.kt
package xyz.speakeasyapp.app.voicefilter.dsp

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * 512-point radix-2 Cooley-Tukey FFT — the lower-latency sibling of
 * [Fft1024]. Identical math; SIZE=512 (SIZE_LOG2=9). Used by
 * [PhaseVocoderPitchShifter] to halve the analysis window (and thus the
 * ~30ms group delay measured at 1024) for the 1.0.x voice-filter latency
 * fix, at the cost of coarser bin resolution (48000/512 ≈ 94 Hz/bin vs
 * 47). Allocation-free per-call after init().
 */
internal class Fft512 {
  companion object {
    const val SIZE = 512
    private const val SIZE_LOG2 = 9
  }

  private val cosTable = FloatArray(SIZE / 2)
  private val sinTable = FloatArray(SIZE / 2)

  init {
    for (i in 0 until SIZE / 2) {
      val theta = -2.0 * PI * i / SIZE
      cosTable[i] = cos(theta).toFloat()
      sinTable[i] = sin(theta).toFloat()
    }
  }

  fun forward(re: FloatArray, im: FloatArray) {
    require(re.size == SIZE && im.size == SIZE) { "buffers must be SIZE=$SIZE" }
    bitReverse(re, im)
    butterflies(re, im)
  }

  fun inverse(re: FloatArray, im: FloatArray) {
    require(re.size == SIZE && im.size == SIZE) { "buffers must be SIZE=$SIZE" }
    for (i in 0 until SIZE) im[i] = -im[i]
    bitReverse(re, im)
    butterflies(re, im)
    val s = 1f / SIZE
    for (i in 0 until SIZE) {
      re[i] *= s
      im[i] = -im[i] * s
    }
  }

  private fun bitReverse(re: FloatArray, im: FloatArray) {
    var j = 0
    for (i in 1 until SIZE) {
      var bit = SIZE ushr 1
      while (j and bit != 0) {
        j = j xor bit
        bit = bit ushr 1
      }
      j = j xor bit
      if (i < j) {
        val tr = re[i]; re[i] = re[j]; re[j] = tr
        val ti = im[i]; im[i] = im[j]; im[j] = ti
      }
    }
  }

  private fun butterflies(re: FloatArray, im: FloatArray) {
    var len = 2
    var stage = 0
    while (stage < SIZE_LOG2) {
      val halfLen = len / 2
      val tableStep = SIZE / len
      var i = 0
      while (i < SIZE) {
        var k = 0
        for (jj in 0 until halfLen) {
          val l = i + jj
          val u = l + halfLen
          val cosK = cosTable[k]
          val sinK = sinTable[k]
          val tr = re[u] * cosK - im[u] * sinK
          val ti = re[u] * sinK + im[u] * cosK
          re[u] = re[l] - tr
          im[u] = im[l] - ti
          re[l] = re[l] + tr
          im[l] = im[l] + ti
          k += tableStep
        }
        i += len
      }
      len = len shl 1
      stage++
    }
  }
}

File: apps/mobile/ios/SpeakeasyBridges/VoiceFilter/dsp/Fft512.swift
//
//  Fft512.swift
//  Speakeasy — Private Call voice filter
//
//  512-point radix-2 Cooley-Tukey FFT — lower-latency sibling of
//  Fft1024 (Swift port of Fft512.kt). Used by PhaseVocoderPitchShifter
//  to halve the analysis window for the 1.0.x latency fix (measured
//  30.3 ms @1024 → 16.6 ms @512), at the cost of coarser bin resolution
//  (48000/512 ≈ 94 Hz/bin). Allocation-free per call after init().
//

import Foundation

final class Fft512 {
  static let SIZE = 512
  private static let SIZE_LOG2 = 9

  private let cosTable: [Float]
  private let sinTable: [Float]

  init() {
    var c = [Float](repeating: 0, count: Fft512.SIZE / 2)
    var s = [Float](repeating: 0, count: Fft512.SIZE / 2)
    for i in 0..<(Fft512.SIZE / 2) {
      let theta = -2.0 * Double.pi * Double(i) / Double(Fft512.SIZE)
      c[i] = Float(cos(theta))
      s[i] = Float(sin(theta))
    }
    self.cosTable = c
    self.sinTable = s
  }

  func forward(_ re: inout [Float], _ im: inout [Float]) {
    precondition(re.count == Fft512.SIZE && im.count == Fft512.SIZE,
                 "buffers must be SIZE=\(Fft512.SIZE)")
    bitReverse(&re, &im)
    butterflies(&re, &im)
  }

  func inverse(_ re: inout [Float], _ im: inout [Float]) {
    precondition(re.count == Fft512.SIZE && im.count == Fft512.SIZE,
                 "buffers must be SIZE=\(Fft512.SIZE)")
    for i in 0..<Fft512.SIZE { im[i] = -im[i] }
    bitReverse(&re, &im)
    butterflies(&re, &im)
    let s = 1.0 / Float(Fft512.SIZE)
    for i in 0..<Fft512.SIZE {
      re[i] *= s
      im[i] = -im[i] * s
    }
  }

  private func bitReverse(_ re: inout [Float], _ im: inout [Float]) {
    var j = 0
    let n = Fft512.SIZE
    for i in 1..<n {
      var bit = n >> 1
      while (j & bit) != 0 {
        j ^= bit
        bit >>= 1
      }
      j ^= bit
      if i < j {
        let tr = re[i]; re[i] = re[j]; re[j] = tr
        let ti = im[i]; im[i] = im[j]; im[j] = ti
      }
    }
  }

  private func butterflies(_ re: inout [Float], _ im: inout [Float]) {
    let n = Fft512.SIZE
    var len = 2
    var stage = 0
    while stage < Fft512.SIZE_LOG2 {
      let halfLen = len / 2
      let tableStep = n / len
      var i = 0
      while i < n {
        var k = 0
        for jj in 0..<halfLen {
          let l = i + jj
          let u = l + halfLen
          let cosK = cosTable[k]
          let sinK = sinTable[k]
          let tr = re[u] * cosK - im[u] * sinK
          let ti = re[u] * sinK + im[u] * cosK
          re[u] = re[l] - tr
          im[u] = im[l] - ti
          re[l] = re[l] + tr
          im[l] = im[l] + ti
          k += tableStep
        }
        i += len
      }
      len <<= 1
      stage += 1
    }
  }
}
